chainstate/src/detail/chainstateref/mod.rs

```diff
@@ -1330,39 +1452,123 @@ impl<'a, S: BlockchainStorageWrite, V: TransactionVerificationStrategy> Chainsta
         self.db_tx.add_block(block).map_err(BlockError::from)
     }
 
-    #[log_error]
-    pub fn set_block_index(&mut self, block_index: &BlockIndex) -> Result<(), BlockError> {
-        self.db_tx.set_block_index(block_index).map_err(BlockError::from)
-    }
-
     #[log_error]
     pub fn set_new_block_index(&mut self, block_index: &BlockIndex) -> Result<(), BlockError> {
         if self.db_tx.get_block_index(block_index.block_id())?.is_some() {
             return Err(BlockError::BlockIndexAlreadyExists(*block_index.block_id()));
         }
-        self.set_block_index(block_index)
+        self.db_tx.set_block_index(block_index)?;
+
+        self.db_tx.mark_as_leaf(block_index, true)?;
+
+        if let Some(non_genesis_parent) =
+            block_index.prev_block_id().classify(self.chain_config).chain_block_id()
+        {
+            let parent_index = self
+                .get_existing_block_index(&non_genesis_parent)
+                .map_err(BlockError::PropertyQueryError)?;
+            self.db_tx.mark_as_leaf(&parent_index, false)?;
+        }
+
+        Ok(())
     }
 
-    /// Delete the block index for the specified block id.
-    /// Panic if the block is marked as persisted.
+    /// Delete the specified block indices.
+    /// Panic if one of the blocks is marked as persisted.
     #[log_error]
-    pub fn del_block_index_of_non_persisted_block(
+    pub fn del_block_indices_of_non_persisted_block_tree<'b>(
```

Comment:
The complexity in this function is terrifying. Do we really need to do these deletions now? Or do we want to wait until we implement that circular buffer solution that doesn't persist them anyway?

Reply:
> The complexity in this function is terrifying.

Just in case, this function was only called from `reset_block_failure_flags`, which can only be manually called via rpc. And the number of blocks to be traversed was supposed to be small.
Anyway, I've reworked `reset_block_failure_flags`, so searching for new leaves doesn't have quadratic complexity anymore.